### PR TITLE
long, jit: rbigint follow-ups — write-barrier and 3.14 surface fixes, and walker specializations for long/int div, mod, pow and divmod

### PR DESCRIPTION
**Pull request by youknowone**
Follow-up work on the ported `rbigint`, stacked on the end-to-end port (#761) and the first review round (#781).

## Correctness

- **`long`/`math`/`gc`** — zero-length signed `int.to_bytes`, `math.lcm` argument validation, and the placement write-barrier fallbacks.
- **`listobject`** — adopt the 3.14 negative-cursor exhausted sentinel for list iterators.
- **`gc`** — apply the write barrier in the blackhole allocator's ref setters. Previously a ref stored through the blackhole path could leave an old-gen object pointing into the nursery without recording the edge.
- **`long`/`gc`** — keep `rbigint` clones fresh and align the wasm old-gen path.
- **`descroperation`** — `divmod` was the only numeric binop missing the
  `needs_numeric_binop_dispatch` override gate, so a subclass overriding
  `__divmod__`/`__rdivmod__` was skipped. This is interpreter-level and
  reproduces with `PYRE_NO_JIT=1`.

## Walker specializations

- `long op int` now takes the `rbigint` `int_*` machine-word legs instead of
  wrapping the `W_IntObject` operand into a bigint first.
- New walker arms for `long // int`, `long % int`, and `long ** int`.
- `UNPACK_SEQUENCE` folds for the object-pair specialised tuple.
- `divmod(int, int)` is specialized on two exact ints. It was previously an
  opaque `bh_call_fn` may-force for *every* operand type: `divmod(a, b)` ran
  3.4x slower than CPython 3.14 while pyre's own `a // b` + `a % b` was 81x
  faster. The arm emits the `makespecialisedtuple2` `Cls_ii` shape directly
  (`specialisedtupleobject.py:169-179`) plus the `ll_int_py_div` /
  `ll_int_py_mod` legs with the `rint.py` `_ovf_zer` domain guards.

  Measured on the int/int loop: **0.405s → 0.004s (101x)**, i.e. from 3.4x
  slower than CPython 3.14 to 29x faster. The optimized loop contains no
  `CallMayForce`, `ForceToken`, or `NewWithVtable`.

## Tests

- `ci` — run `pyre/extra_tests/parity_tests` in the `pyre/check.py` jobs; they
  were not previously gated.
- New parity tests: `long_int_div_mod_jit.py`, `long_int_pow_jit.py`,
  `rbigint_mixed_int_jit.py`, `unpack_specialised_pair_shapes.py`,
  `list_iterator_surface.py`, `math_lcm_argument_validation.py`, plus
  `divmod_int_jit.py`.
- `rbigint_differential.rs` gains the `int_*` family.

## Verification

Measured at this branch tip with a freshly extracted `build/llbc` census and
**both** backend binaries rebuilt against it, using `python3.14` as the oracle
(`PYRE_CHECK_PYTHON3=python3.14`):

| test | dynasm | cranelift | without JIT | pre-existing? |
|---|---|---|---|---|
| `long_int_div_mod_jit.py` | FAIL | OK | OK | **no — added by this branch** |
| `bytearray_python314.py` | FAIL | FAIL | FAIL (139) | yes (#560) |
| `memoryview_python314.py` | FAIL | FAIL | FAIL (139) | yes (#560) |
| `range_python314.py` | FAIL | FAIL | FAIL (1) | yes (#560) |
| `slice_python314.py` | FAIL | FAIL | FAIL (1) | yes (#560) |
| `set_iterator_python314.py` | FAIL | FAIL | FAIL (1) | yes (#629) |

Everything else in the parity suite is green on both backends.

An earlier `check.py --backend dynasm,cranelift` → 326/326 run **should be
disregarded**: it was measured against a stale `build/llbc/*.ullbc`. `cargo build`
does not re-extract LLBC, so that run validated traces built from a census
extracted hours earlier from older sources.

### The five pre-existing failures

These are not regressions from this branch. All five fail identically on both
backends *and* under `PYRE_NO_JIT=1`, so they are interpreter-level, and the
tests were added in #560/#629 long before this work. They surface here only
because this branch's `ci: run pyre/extra_tests/parity_tests in the
pyre/check.py jobs` commit gates them for the first time.

`range`/`slice`/`set_iterator` are the same shape — calling an unbound method
descriptor with no argument reports `TypeError: count() missing 1 required
positional argument` instead of CPython's `unbound method ... needs an
argument`. `bytearray`/`memoryview` segfault (139).

They need fixing (or the gate needs to land separately), but they are
independent of the rbigint work.

## Branch defect — blocks merge

`long_int_div_mod_jit.py`, added by this branch, aborts under `pyre-dynasm`:

```
GC BUG: invalid type_id=... at obj_addr=... site=copy_nursery_object
  MiniMarkGC::copy_nursery_object <- trace_and_update_object
  <- do_collect_nursery <- alloc_with_type <- dynasm_nursery_slowpath
```

**Confirmed present at this branch tip** by a control run (committed tip, clean
working tree, LLBC re-extracted so the census matches the sources): 6/6. It was
masked earlier only by the stale census.

### Characterisation

- **dynasm only** — cranelift passes, consistent with `dynasm_nursery_slowpath`
  on the stack. (Note cranelift's shim never collects at allocation sites, so it
  simply never exercises this path; dynasm is the orthodox one here.)
- **JIT-only** — passes under `PYRE_NO_JIT=1`.
- **Cumulative and geometry-dependent** — the failing function passes in
  isolation; it needs ~5 earlier allocation-heavy functions. Shrinking the
  nursery (`PYPY_GC_NURSERY` 128K–512K) does **not** reproduce it; it depends on
  the default 4 MB fill geometry.

### Root-cause analysis so far

An instrumented collector (probe naming the *referrer*, skipping forwarded
headers) gives:

```
bad child via varsize_item  parent_type_id=9 (PY_OBJECT_ARRAY) parent=OLD-GEN
capacity=8, item[0]=<nursery addr>  BAD-TID
item[1..7]=<low nursery addrs>      valid
child FORWARDED_EARLIER_TO=None
```

The parent is an old-gen `ItemsBlock`; only `item[0]` is corrupt. It holds the
*pre-collection* address of an object that a forwarding-history probe shows was
**never forwarded by any collection in the whole run** — i.e. the object was
unreachable from every GC root when its collection ran, yet compiled code kept
using its address afterwards and stored it into a freshly allocated block.

**So this is a missing-GC-root defect, not a write-barrier or zeroing defect.**
Some holder keeps a `PyObjectRef` across a collection without the GC knowing.

### Hypotheses refuted with evidence (recorded so they are not re-tried)

| hypothesis | how it was refuted |
|---|---|
| `malloc_zero_filled` misconfigured | both backends set it `false`, citing incminimark.py:211. Only the doc comment at `rewrite.rs:140-150` is stale (it claims `Nursery::reset()` zeroes; the real `reset()` zeroes on wasm32 only) |
| uninitialized capacity tail | allocation helpers NULL-fill `len..cap`; poison mode would raise the dedicated poison panic instead |
| `wb_covered` drops a needed barrier | dumped compiled code: `CondCallGcWbArray` is emitted before every `GcStoreIndexed` on the append arm |
| born-old-gen block missing TRACK_YOUNG_PTRS | `finish_alloc_in_oldgen` sets it; and the block is 72 bytes, far below `large_object_threshold` |
| card marking incoherent | `HAS_CARDS` is only ever set by tests, so production degenerates to whole-object remembering (correct) |
| **old-gen spill from `alloc_with_type` (`collector.rs:868-871`) breaking the rewrite's WB elision** | **instrumented: ZERO spills in a failing run.** This was the leading hypothesis; it is empirically dead for this crash |
| malloc slow-path register handling | x2–x13/x19–x22 are spilled and reloaded from GC-updated slots; regalloc matches aarch64/regalloc.py:958 (pyre spills *more*) |
| baked Ref constants unpinned | constants resolve through a traced `GcTable` (`gcreftracer.rs`, the RPython GCREFTRACER shape) |

### Still open

Identifying the specific unrooted holder. The diagnostic patch used above is
kept out of tree.

Separately, two genuine latent divergences were found and are **not** the cause
here, but should be fixed on their own:

1. `alloc_with_type` can return old-gen memory where incminimark's
   `collect_and_reserve` (incminimark.py:865-930) loops minor collections and
   never does. The rewrite's write-barrier elision for fresh mallocs
   (rewrite.py:911) depends on that young-result guarantee.
2. `emit_write_barrier_fastpath_kind` silently emits nothing when the WB descr
   is unavailable (`_ => return`, both arches), and the aarch64 inline card
   write silently skips non-register indices. Upstream asserts in both cases.

— *authored by Claude*

## Summary by CodeRabbit

* **New Features**
  * Added CI parity step running extra parity tests under the same CPython interpreter.
  * Expanded parity coverage (JIT divmod, long div/mod, long pow, unpacking, conversions/edge cases) and added clear completion output.
* **Bug Fixes**
  * Improved write-barrier descriptor handling, blackhole write barriers, and GC/object alignment.
  * Strengthened constant loading/identity and integer parsing/formatting behaviors.
  * Enhanced iterator exhaustion/restoration and range/list cursor handling.
* **Tests**
  * Added GC-poison option and broadened JIT/GC/translation/numeric differential assertions.

### CODE REVIEW

File: pyre/pyre-interpreter/src/eval.rs [hunk 1]
```diff
@@ -272,6 +272,11 @@ pub unsafe fn walk_raw_code_roots(
         }
         let code = &mut *(value as *mut crate::pycode::PyCode);
         visitor(&mut *(&mut code.w_globals as *mut PyObjectRef as *mut majit_ir::GcRef));
+        if !code.co_consts_w.is_null() {
+            for slot in (&mut *code.co_consts_w).iter_mut() {
+                visitor(&mut *(slot as *mut PyObjectRef as *mut majit_ir::GcRef));
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/eval.rs]: Fixed by walking nested code constants through the live-code root traversal; the final implementation also registers every Box-immortal PyCode wrapper so standalone code objects are covered.

File: pyre/pyre-interpreter/src/module/_io/textio.rs [hunk 2]
```diff
@@ -1138,21 +1143,11 @@ impl W_TextIOWrapper {
...
-            return Ok(w_int_new(0));
-        }
         let input = snapshot.input.clone();
-        let mut cookie = PositionCookie {
-            start_pos: raw_pos - input.len() as u64,
-            dec_flags: snapshot.flags,
-            ..PositionCookie::default()
-        };
+        cookie.dec_flags = snapshot.flags;
+        cookie.start_pos = cookie.start_pos.wrapping_sub(input.len() as u64);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/module/_io/textio.rs]: Fixed: tell() now returns zero before subtracting when the snapshot input is larger than the reported raw position, matching the upstream guard.

File: majit/majit-translate/src/translator/rtyper/flowspace_adapter.rs [hunk 3]
```diff
@@ -804,10 +858,10 @@ pub(crate) fn op_canraise(kind: &OpKind) -> bool {
         // the 0-arg unit-variant ctor, which `translate_op` pre-folds to a
         // `Constant` and emits no op — `op_canraise` is false exactly when
         // that happens.  Matched before the general `Call` arm.
-        OpKind::Call {
+        kind @ OpKind::Call {
             target: crate::model::CallTarget::SyntheticTransparentCtor { .. },
             ..
-        } => !is_elided_unit_variant_ctor(kind),
+        } => !is_elided_unit_variant_ctor(kind) && !is_array_aggregate_ctor(kind),
```
[the remark this answers is not in the document]
Comment (youknowone) [reply majit/majit-translate/src/translator/rtyper/flowspace_adapter.rs]: Fixed by sharing the complete aggregate-rewrite decision between translation and op_canraise, including the element-extraction requirements, so partial recognition no longer disagrees.

File: majit/majit-translate/src/front/rbigint_call.rs [hunk 4]
```diff
@@ -180,23 +206,27 @@ pub(crate) fn int_comparison_residual_for_method(leaf: &str) -> Option<Vec<Strin
...
-            "descroperation",
-            "jit_bigint_pow_nomod",
-        ]
-        .into_iter()
-        .map(str::to_string)
-        .collect(),
+        ["pyre_interpreter", "objspace", "descroperation", residual]
+            .into_iter()
+            .map(str::to_string)
+            .collect(),
```
[the remark this answers is not in the document]
Comment (youknowone) [reply majit/majit-translate/src/front/rbigint_call.rs:229]: Fixed in c909b068f3: both bigint_int_pow_nomod and bigint_int_floordiv_nonzero now have direct source-to-residual mapping assertions.

File: pyre/pyre-interpreter/src/objspace/descroperation.rs [hunk 5]
```diff
@@ -723,89 +872,222 @@ unsafe fn long_add(a: PyObjectRef, b: PyObjectRef) -> PyResult {
...
+    };
+    if is_long(a) && va.get_sign() == 1 && vb.int_eq(1) {
+        return Ok(pyre_object::longobject::w_long_from_raw(
+            w_long_get_raw_value(a),
+        ));
+    }
     // rbigint.floordiv → _divmod, returning the quotient half (rbigint.py:1001).
     // `_floordiv`/`_int_floordiv` both `newlong` the quotient, keeping a long.
-    Ok(w_long_new(bigint_floordiv_nonzero(as_bigint(a), vb)))
+    Ok(w_long_new(bigint_floordiv_nonzero(va, vb)))
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/objspace/descroperation.rs]: Fixed in bb34efbdfe: removed the non-upstream payload aliases for long // long(1) and the divmod quotient, retaining only the upstream-valid long/int fast path.

File: pyre/pyre-interpreter/src/objspace/std/formatting.rs [hunk 6]
```diff
@@ -241,7 +241,21 @@ fn cformat_rbigint(spec: &CFormatSpec, num: &BigInt) -> String {
...
+    };
+    let negative = num.int_lt(0);
+    let mut magnitude = num.format(digits, "", "", 0).map_err(|error| match error {
+        pyre_object::rbigint::RBigIntError::Memory => PyError::memory_error(""),
+        _ => unreachable!("validated radix formatting returned an unrelated error"),
+    })?;
+    if negative {
+        debug_assert!(magnitude.starts_with('-'));
+        magnitude.remove(0);
+    }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/objspace/std/formatting.rs:262]: Fixed in bb34efbdfe: sign removal uses strip_prefix and reports a SystemError if the bigint formatting invariant is violated in any build profile.

File: pyre/pyre-interpreter/src/pycode.rs [hunk 7]
```diff
@@ -436,6 +447,46 @@ fn box_code_constant_with_firstlineno(code: &crate::CodeObject, firstlineno: i32
...
+    }
+    let slots = unsafe { &*dst_code.co_consts_w };
+    for (index, slot) in slots.iter().enumerate() {
+        let value = unsafe { w_code_const(src, index) };
+        if !value.is_null() {
+            slot.store(value, std::sync::atomic::Ordering::Release);
+        }
+    }
+    pyre_object::gc_roots::mark_prebuilt_roots_dirty();
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/pycode.rs:540]: Fixed in bb34efbdfe: code replacement copies only already-realized atomic slots and does not realize missing source constants.

File: pyre/pyre-object/src/rbigint.rs [hunk 8]
```diff
@@ -4284,6 +4472,38 @@ struct PartsCacheBase {
...
+    }
+}
+
+impl Drop for PendingPartsCacheDigitRoot {
+    fn drop(&mut self) {
+        if self.registered {
+            crate::gc_hook::try_gc_remove_root(self.slot);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-object/src/rbigint.rs:4505]: Fixed in bb34efbdfe: the digit field address is formed with addr_of_mut! from the raw Arc pointer without materializing an aliased mutable reference.

File: pyre/pyre-object/tests/rbigint_differential.rs [hunk 9]
```diff
@@ -69,6 +69,24 @@ fn oracle_floor_divmod(a: &Oracle, b: &Oracle) -> (Oracle, Oracle) {
...
+        1 => 1,
+        2 => -1,
+        3 => 1_i64 << (raw >> 32) % 62,
+        4 => -(1_i64 << (raw >> 32) % 62),
+        5 => i64::MAX,
+        6 => i64::MIN,
+        _ => raw as i64,
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-object/tests/rbigint_differential.rs]: Fixed in bb34efbdfe: the boundary/random-arm documentation and shift parentheses now match the generator.

File: pyre/pyre-object/tests/rbigint_differential.rs [hunk 10]
```diff
@@ -230,6 +248,72 @@ fn deterministic_arbitrary_size_arithmetic_matches_independent_oracle() {
...
+            );
+            let (iq, ir) = a.int_divmod(word).unwrap();
+            assert_eq!(to_oracle(&iq), oq, "int_divmod q, iteration {iteration}");
+            assert_eq!(to_oracle(&ir), or, "int_divmod r, iteration {iteration}");
+        } else {
+            assert!(a.int_floordiv(word).is_err());
+            assert!(a.int_mod(word).is_err());
+            assert!(a.int_mod_int_result(word).is_err());
+            assert!(a.int_divmod(word).is_err());
+        }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-object/tests/rbigint_differential.rs:327]: Fixed in bb34efbdfe: all zero-divisor assertions include the iteration context.

File: pyre/pyre-interpreter/src/pycode.rs [hunk 11]
```diff
@@ -1360,22 +1417,42 @@ pub unsafe fn w_code_co_const(w_code_obj: PyObjectRef, idx: usize) -> PyObjectRe
...
+    let registered = unsafe { pyre_object::gc_hook::try_gc_add_root(candidate_root) };
+    let published = match slot.compare_exchange(
+        std::ptr::null_mut(),
+        realized,
+        std::sync::atomic::Ordering::AcqRel,
+        std::sync::atomic::Ordering::Acquire,
+    ) {
+        Ok(_) => {
+            pyre_object::gc_roots::mark_prebuilt_roots_dirty();
             realized
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/pycode.rs:1500]: Fixed in bb34efbdfe: every live Box-immortal PyCode wrapper is registered in a process-global owner, and the prebuilt-root walker traces standalone code globals and realized constant slots. Added a regression test for standalone root visibility.

File: pyre/pyre-jit-trace/src/jitcode_dispatch/residual_call.rs [hunk 12]
```diff
@@ -3873,6 +3887,23 @@ pub(crate) fn dispatch_residual_call_iIRd_kind<Sym: WalkSym>(
...
                             )?;
                         }
+                        if specialized.is_none() {
+                            // `_int_floordiv` / `_int_mod` are the same family:
+                            // an Int divisor keeps its machine word instead of
+                            // being widened to a bigint, and `_int_mod`'s
+                            // result is a machine int rather than a long.
+                            specialized = try_walker_specialize_binary_op_long_int_div(
+                                ctx, op.pc, op_tag, &r_args, &allboxes, call_descr, dst, dst_bank,
+                            )?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-jit-trace/src/jitcode_dispatch/residual_call.rs:4155]: Fixed the underlying GC-root/resume defect rather than disabling the long/int specialization. With freshly extracted LLBC, the full 85-script parity suite passes under both dynasm and cranelift with GC poisoning, including long_int_div_mod_jit.py.

File: .github/workflows/pyre-ci.yml [hunk 13]
```diff
@@ -383,6 +383,15 @@ jobs:
...
       run: ${{ steps.cpython.outputs.python-path }} pyre/check.py
+    - name: Run pyre/extra_tests/parity_tests
+      # Every script must exit 0 and print "OK" as its last line under CPython
+      # AND under each pyre backend, so a divergence from CPython observable
+      # semantics fails here. check.py has just built the release pyre-dynasm
+      # and pyre-cranelift binaries the runner drives, so this costs only the
+      # per-script process spawns.
+      env:
+        PYRE_CHECK_PYTHON3: ${{ steps.cpython.outputs.python-path }}
+      run: ${{ steps.cpython.outputs.python-path }} pyre/extra_tests/parity_tests/run.py
```
[the remark this answers is not in the document]
Comment (youknowone) [reply .github/workflows/pyre-ci.yml:394]: The parity suite is no longer already-failing: all 85 scripts pass locally under CPython, dynasm, and cranelift with freshly extracted LLBC and GC poisoning. Keeping this as a mandatory gate is intentional.

[the remark this answers is not in the document]
Comment (youknowone) [reply .github/workflows/pyre-ci.yml:394]: The missing-root dynasm failure has been fixed at the runtime/root-walking layer; the full parity gate now passes under dynasm and cranelift. The test remains enabled and the gate remains mandatory.

File: majit/majit-backend/src/resume_value.rs [hunk 14]
```diff
@@ -46,6 +54,9 @@ impl ResumeValueSource {
             ResumeValueSource::FailArg(_) => ResumeValueKind::FailArg,
             ResumeValueSource::Constant(_) => ResumeValueKind::Constant,
             ResumeValueSource::Virtual(_) => ResumeValueKind::Virtual,
+            ResumeValueSource::Tagged(_) => {
+                panic!("runtime tagged resume source has no exported ResumeValueKind")
+            }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply majit/majit-backend/src/resume_value.rs:59]: Verified as an intentional representation invariant. Tagged values are the raw RPython fieldnums shape and are consumed only by ResumeDataDirectReader, which decodes ref/int/float through the live forwarded rd_consts pool. Layout/export paths construct structured variants. The panic arms reject an illegal leak across that boundary; they are not reachable consumers of rd_virtual_to_virtual_info output.

File: majit/majit-metainterp/src/pyjitpl.rs [hunk 15]
```diff
@@ -1666,29 +1666,56 @@ impl<M: Clone> MetaInterp<M> {
...
-            let consts = unsafe { s.rd_consts_mut_for_gc() };
+            let consts = unsafe { pool.as_mut_vec_for_gc() };
             for c in consts.iter_mut() {
                 if let Const::Ref(slot) = c {
                     visitor(slot);
                 }
             }
         }
 
+        let mut visited = Vec::new();
```
[the remark this answers is not in the document]
Comment (youknowone) [reply majit/majit-metainterp/src/pyjitpl.rs:1691]: Keeping the transient Vec intentionally. RPython relies on collector mark state and has no HashSet side table here; pyre only needs a small identity list to compensate for shared Arc pools during one root walk. Introducing a non-upstream HashSet would violate the port data-structure parity rule without evidence that this traversal is a bottleneck.

File: pyre/pyre-interpreter/src/builtins.rs [hunk 16]
```diff
@@ -12068,30 +12072,29 @@ pub(crate) fn builtin_round(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::
...
+                w_long_get_value(obj)
+            } else {
+                owned_a = BigInt::from(w_int_get_value(obj));
+                &owned_a
+            };
+            let exponent = nd.neg();
+            let b = BigInt::from(10)
+                .pow(&exponent, None)
+                .map_err(|_| crate::PyError::memory_error("exponent too large"))?;
             // `_PyLong_DivmodNear`: q = round(a / b) ties-to-even,
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/builtins.rs:12180]: Keeping the upstream algorithm: pypy/objspace/std/intobject.py descr_round literally computes rbigint.fromint(10).pow(ndigits.neg()) before divmod_near. A bit-length shortcut would be a non-upstream optimization that changes allocation/error timing, so strict PyPy structural parity takes precedence here.

File: pyre/pyre-interpreter/src/eval.rs [hunk 17]
```diff
@@ -266,12 +266,48 @@ pub unsafe fn walk_raw_code_roots(
...
+    }
+
     unsafe {
         if value.is_null() || !crate::pycode::is_code(value) {
             return;
         }
-        let code = &mut *(value as *mut crate::pycode::PyCode);
-        visitor(&mut *(&mut code.w_globals as *mut PyObjectRef as *mut majit_ir::GcRef));
+        walk(value, visitor, &mut Vec::new());
     }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/eval.rs:314]: Keeping the transient Vec intentionally. Upstream PyPy gets transitive cycle suppression from GC mark state, not a BTreeSet/HashSet side table. Because Box-immortal PyCode wrappers sit outside pyre GC, this small per-walk identity Vec is the closest compensation and preserves the required upstream ownership shape.
